src/internal_rep.rs

```diff
@@ -1539,6 +1541,121 @@ fn call_to_fc_rules<'a>(
     Ok(ret)
 }
 
+#[derive(Clone, Copy, Debug, Eq, PartialEq, PartialOrd, Ord)]
+pub enum Protocol {
+    Tcp,
+    Udp,
+}
+
+impl From<Protocol> for &str {
+    fn from(p: Protocol) -> &'static str {
+        match p {
+            Protocol::Tcp => "tcp",
+            Protocol::Udp => "udp",
+        }
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, PartialOrd, Ord)]
+pub struct PortconRule<'a> {
+    proto: Protocol,
+    port: CascadeString, // TODO: lists and ranges
+    context: Context<'a>,
+}
+
+impl From<&PortconRule<'_>> for sexp::Sexp {
+    fn from(p: &PortconRule) -> sexp::Sexp {
+        list(&[
+            atom_s("portcon"),
+            atom_s(p.proto.into()),
+            Sexp::Atom(Atom::S(p.port.to_string())),
+            Sexp::from(&p.context),
+        ])
+    }
+}
+
+pub fn call_to_portcon_rule<'a>(
+    c: &FuncCall,
+    types: &TypeMap,
+    class_perms: &ClassList,
+    context: &BlockContext,
+    file: &SimpleFile<String, String>,
+    parent_type: &'a TypeInfo,
+) -> Result<PortconRule<'a>, CascadeErrors> {
+    let target_args = vec![
+        FunctionArgument::new(
+            &DeclaredArgument {
+                param_type: CascadeString::from("string"),
+                is_list_param: false,
+                name: CascadeString::from("protocol"),
+                default: None,
+            },
+            types,
+            None,
+        )?,
+        FunctionArgument::new(
+            &DeclaredArgument {
+                param_type: CascadeString::from("number"),
+                is_list_param: false, // TODO: Need to support lists and ranges
+                name: CascadeString::from("port"),
+                default: None,
+            },
+            types,
+            None,
+        )?,
+    ];
+
+    let validated_args =
+        validate_arguments(c, &target_args, types, class_perms, context, Some(file))?;
+    let mut args_iter = validated_args.iter();
+
+    let proto = args_iter
+        .next()
+        .ok_or_else(|| ErrorItem::Internal(InternalError::new()))?
+        .get_name_or_string(context)?;
+    let proto = match proto.as_ref() {
+        "\"tcp\"" => Protocol::Tcp,
+        "\"udp\"" => Protocol::Udp,
```

Comment:
It may be nice to lowercase the proto string to support stuff like TCP or Tcp

Reply:
My inclination is to say that we don't really want to support really perverse stuff like "tCP" or something.  Feels like that would make the policy less maintainable.  Maybe just enumerating "tcp" and "TCP" would be better than the case insensitive compare?

Reply:
I am good with just TCP or tcp.